actors/builtin/miner/miner_actor.go

```diff
@@ -1417,146 +1708,47 @@ func validateExpiration(rt Runtime, st *State, activation, expiration abi.ChainE
-	var expiredEpochs []abi.ChainEpoch
-	var expiredSectors []*abi.BitField
-	errDone := fmt.Errorf("done")
-	err := st.ForEachSectorExpiration(store, func(expiry abi.ChainEpoch, sectors *abi.BitField) error {
-		if expiry > epoch {
-			return errDone
-		}
-		expiredSectors = append(expiredSectors, sectors)
-		expiredEpochs = append(expiredEpochs, expiry)
-		return nil
-	})
-	if err != nil && err != errDone {
-		return nil, err
-	}
-	err = st.ClearSectorExpirations(store, expiredEpochs...)
-	if err != nil {
-		return nil, fmt.Errorf("failed to clear sector expirations %s: %w", expiredEpochs, err)
+			params.ReplaceSectorNumber, replaceSector.Expiration, params.Expiration)
 	}
 
-	allExpiries, err := bitfield.MultiMerge(expiredSectors...)
-	if err != nil {
-		return nil, fmt.Errorf("failed to union expired sectors: %w", err)
-	}
-	return allExpiries, err
-}
-
-// Removes and returns sector numbers that were faulty at or before an epoch, and returns the sector
-// numbers for other ongoing faults.
-func popExpiredFaults(st *State, store adt.Store, latestTermination abi.ChainEpoch) (*abi.BitField, *abi.BitField, error) {
-	var expiredEpochs []abi.ChainEpoch
-	var expiredFaults []*abi.BitField
-	var ongoingFaults []*abi.BitField
-	errDone := fmt.Errorf("done")
-	err := st.ForEachFaultEpoch(store, func(faultStart abi.ChainEpoch, faults *abi.BitField) error {
-		if faultStart <= latestTermination {
-			expiredFaults = append(expiredFaults, faults)
-			expiredEpochs = append(expiredEpochs, faultStart)
-		} else {
-			ongoingFaults = append(ongoingFaults, faults)
-		}
-		return nil
-	})
-	if err != nil && err != errDone {
-		return nil, nil, nil
-	}
-	err = st.ClearFaultEpochs(store, expiredEpochs...)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to clear fault epochs %s: %w", expiredEpochs, err)
-	}
+	status, err := st.SectorStatus(store, params.ReplaceSectorDeadline, params.ReplaceSectorPartition, params.ReplaceSectorNumber)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to check sector health %v", params.ReplaceSectorNumber)
 
-	allExpiries, err := bitfield.MultiMerge(expiredFaults...)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to union expired faults: %w", err)
-	}
-	allOngoing, err := bitfield.MultiMerge(ongoingFaults...)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to union ongoing faults: %w", err)
+	switch status {
+	case SectorNotFound:
+		rt.Abortf(exitcode.ErrIllegalArgument, "sector %d not found at %d:%d (deadline:partition)",
+			params.ReplaceSectorNumber, params.ReplaceSectorDeadline, params.ReplaceSectorPartition,
+		)
+	case SectorFaulty:
+		rt.Abortf(exitcode.ErrIllegalArgument, "cannot replace faulty sector %d", params.ReplaceSectorNumber)
+	case SectorTerminated:
+		rt.Abortf(exitcode.ErrIllegalArgument, "cannot replace terminated sector %d", params.ReplaceSectorNumber)
+	case SectorHealthy:
+		// pass
+	default:
+		// TODO: Just panic?
```

Comment:
Since the exits are all IllegalArgument, return an error instead of status enum?

Reply:
I'm returning an error when we have an illegal state. I could return two errors...

Reply:
Right, just panic (`Assert`) for now. The exit code error could help later.